sway/ipc-server.c

```diff
@@ -349,43 +349,53 @@ static int ipc_selection_data_cb(int fd, uint32_t mask, void *data) {
 	}
 
 	if (mask & WLC_EVENT_READABLE) {
-		static const int step_size = 512;
-		char *data = NULL;
-		int ret = 0;
-		int current = 0;
+		static const int max_size = 8192 * 1000;
+		int len = 512;
+		int i = 0;
+		char *buf = malloc(len);
 
 		// read data as long as there is data avilable
 		// grow the buffer step_size in every iteration
-		do {
-			if (data == NULL) {
-				data = malloc(step_size);
-			} else {
-				data = realloc(data, current + step_size);
-			}
+		for(;;) {
+			int amt = read(fd, buf + i, len - i - 1);
+			if (amt <= 0)
+				break;
+
+			i += amt;
+			if (i >= len - 1) {
+				if (len >= max_size) {
+					sway_log(L_ERROR, "selection data too large");
+					free(buf);
+					goto cleanup;
+				}
+				char *next = realloc(buf, (len *= 2));
+				if (!next) {
+					sway_log_errno(L_ERROR, "relloc failed");
+					free(buf);
+					goto cleanup;
+				}
```

Comment:
Your error messages could be more descriptive. L_ERROR is bubbled up to the user by default, they should be able to understand them.